sematic/resolvers/state_machine_resolver.py

```diff
@@ -92,6 +110,15 @@ def _run_inline(self, future: AbstractFuture):
     def _wait_for_scheduled_run(self) -> None:
         pass
 
+    def _handle_sigint_cancel(self, signum, frame):
+        logger.warning("Received SIGINT, canceling resolution...")
+        self._resolution_did_cancel()
+
+    def _cancel_futures(self):
+        for future in self._futures:
+            if not future.state.is_terminal():
+                self._set_future_state(future, FutureState.CANCELED)
```

Comment:
This is actually inconsistent with our goal to make the server the "source of truth" for state information. If the server fails to actually cancel a run for some reason, the in-memory future will be `CANCELED` but the run will not be. Maybe make `_cancel_futures` an abstract method, have the `SilentResolver` do this, the `LocalResolver` do this and update the server, and the cloud resolver just ask the server to cancel and poll for updates on the statuses

Reply:
Runs are already canceled on the server-side at this point. The purpose of this is to make the resolution loop exits gracefully. We could not even do this at all and be fine since we do `exit(1)`.

Reply:
I would prefer something more like this:

```python
    def _cancel_non_terminal_futures(self):
        self._refresh_future_states()
        for future in self._futures:
            if not future.state.is_terminal():
                raise RuntimeError(f"Future {future.id} should have been cancelled, but was: {future.state}")
```

But I'll approve and we can revisit in-memory/server state synchronization later